modules/javafx.graphics/src/main/java/com/sun/javafx/css/ImmutablePseudoClassSetsCache.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.sun.javafx.css;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import javafx.css.PseudoClass;
+
+/**
+ * A cache for immutable sets of {@link PseudoClass}es.
+ */
+public class ImmutablePseudoClassSetsCache {
+    private static final Map<Set<PseudoClass>, Set<PseudoClass>> CACHE = new HashMap<>();
+
+    /**
+     * Returns an immutable set of {@link PseudoClass}es.
+     * <p>
+     * Note: this method may or may not return the same instance for the same set of
+     * {@link PseudoClass}es.
+     *
+     * @param pseudoClasses a set of {@link PseudoClass} to make immutable, cannot be {@code null}
+     * @return an immutable set of {@link PseudoClass}es, never {@code null}
+     * @throws NullPointerException when {@code pseudoClasses} is {@code null} or contains {@code null}s
+     */
+    public static Set<PseudoClass> of(Set<PseudoClass> pseudoClasses) {
+        Set<PseudoClass> cachedSet = CACHE.get(Objects.requireNonNull(pseudoClasses, "pseudoClasses cannot be null"));
+
+        if (cachedSet != null) {
+            return cachedSet;
+        }
+
+        Set<PseudoClass> copy = Set.copyOf(pseudoClasses);
+
+        CACHE.put(copy, copy);
+
+        return copy;
```

Comment:
Almost, but the key is also the copied set.  In your version the key is the original object, which may be a modifiable set.  If the key is modified, the cache will not function correctly anymore.

I had a different version before, but @mstr2 pointed out that there was an unnecessary allocation in that version which this version avoids.

Reply:
No, a `Set` wouldn't work here, for this reason:

      if (CACHE.contains(mutableSet)) {
             return ... // how do I get the immutable set from CACHE when it is a Set?
      }

Reply:
I just finished my review of the code, and this (why a `Map` rather than a `Set`) was the only question I had. I looked at the resolved threads, and I see that it has been asked and answered. Good.